Rich Text Editor | Added HardBreak extension in editor for force line break (<br> tag)

aagash-ni opened the pull request:
# Pull Request

## 🤨 Rationale

Fix for [Bug 2516897](https://dev.azure.com/ni/DevCentral/_workitems/edit/2516897): Switching to a new line by pressing enter adds more spaces than expected.

- Pressing enter in the editor for a new line adds a `Paragraph tag` which has more space than expected between lines.
-  Enabling [HardBreak ](https://tiptap.dev/api/nodes/hard-break) will allow the user to switch to a new line by pressing `Shift/Ctrl/Cmd + Enter` which will make minimal space between lines by adding `<br> tag`  instead of `Paragraph Tag`.

## 👩‍💻 Implementation

- Added [`HardBreak` ](https://tiptap.dev/api/nodes/hard-break) node in Tiptap Editor.
- Enabled [`newline`](https://github.com/markdown-it/markdown-it/blob/2b6cac25823af011ff3bc7628bc9b06e483c5a08/lib/rules_inline/newline.js#L27C8-L28C42) and [`escape`](https://github.com/markdown-it/markdown-it/blob/2b6cac25823af011ff3bc7628bc9b06e483c5a08/lib/rules_inline/escape.js#L1) rules in MarkdownIt for `RichTextMarkdownParser` to parse the HardBreak markdown syntax as per [CommonMark Spec](https://spec.commonmark.org/0.30/#hard-line-breaks) to `<br> tag`.
- Added `hardBreak` node to `RichTextMarkdownSerializer` to serialize the `<br> tag` to respective hard break markdown syntax.
- As we are using `prosemirror-markdown`, it serializes the `<br>` tag to backslash with [line ending](https://spec.commonmark.org/0.30/#hard-line-breaks) syntax as per the [CommonMark Spec example 633](https://spec.commonmark.org/0.30/#example-633).

In Windows `Shift/Ctrl+ Enter` and in Mac `Cmd + Enter` will switch to a new line (line break).

We are not exposing any format buttons to add line breaks (Hard Break), The only way to break the line is by using the Keys mentioned above.



## 🧪 Testing

- Added unit tests and visual tests for the functionality.
- Manually tested and verified the functionality of the supported features in Storybook build.

## ✅ Checklist

<!--- Review the list and put an x in the boxes that apply or ~~strike through~~ around items that don't (along with an explanation). -->

- [x] I have updated the project documentation to reflect my changes or determined no changes are needed.

vikisekarNI commented on packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:
```diff
@@ -344,4 +344,52 @@ describe('Markdown parser', () => {
+            tags: string[]
+        }[] = [
+            {
+                name: '**bold**\\\n*Italics*',
+                tags: ['P', 'STRONG', 'BR', 'EM']
```
Can we replace name here as `value` and give a valid test case title with `name` which is easy to understand?

vikisekarNI replied:
For ex: 
1. `should render br tag between bold and italics`
2. `should render br tag between two first level bullet list items`

aagash-ni replied:
Updated

vikisekarNI commented on packages/nimble-components/src/rich-text/models/tests/markdown-serializer.spec.ts:
```diff
@@ -258,4 +255,71 @@ describe('Markdown serializer', () => {
+        for (const value of supportedNodesMarks) {
+            const specType = getSpecTypeByNamedList(value, focused, disabled);
+            specType(
+                `Should return ${value.name} markdown (${value.markdown}) when its respective node is passed`,
```
```suggestion
                `Should serialize ${value.name} to markdown`,
```

aagash-ni replied:
Updated

vikisekarNI commented on packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:
```diff
@@ -335,6 +344,20 @@ describe('RichTextEditor', () => {
+            await pageObject.clickFooterButton(ToolbarButton.bold);
+            await pageObject.setEditorTextContent('bold1');
+            await pageObject.pressShiftEnterKeysInEditor();
+            await pageObject.setEditorTextContent('bold2');
```
```suggestion
            await pageObject.setEditorTextContent('bold after hard break');
```

aagash-ni replied:
Updated

vikisekarNI commented on packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:
```diff
@@ -343,6 +366,20 @@ describe('RichTextEditor', () => {
+            await pageObject.clickFooterButton(ToolbarButton.italics);
+            await pageObject.setEditorTextContent('italics1');
+            await pageObject.pressShiftEnterKeysInEditor();
+            await pageObject.setEditorTextContent('italics2');
```
```suggestion
            await pageObject.setEditorTextContent('italics after hard break');
```

vikisekarNI replied:
nit: Apply similarly for other test cases for a slightly enhanced readability

aagash-ni replied:
Updated as per your suggestion.

vikisekarNI commented on packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:
```diff
@@ -890,6 +1038,39 @@ describe('RichTextEditor', () => {
         }
     });
 
+    describe('Should return markdown with back slash (HardBreak) when markdown with hard break syntax are assigned', () => {
```
Just curious to understand why you wanted to have this test case now? 

Title suggestion
```suggestion
    describe('`getMarkdown` with hard break backslashes should be same immediately after `setMarkdown`', () => {
```

aagash-ni replied:
Updated

vikisekarNI commented on packages/nimble-components/src/rich-text/models/markdown-parser.ts:
```diff
@@ -43,7 +43,9 @@ export class RichTextMarkdownParser {
             'list',
-            'autolink'
+            'escape',
+            'autolink',
```
Never mind. My bad. I missed to see the autolink entry above

aagash-ni commented on packages/nimble-components/src/rich-text/models/markdown-serializer.ts:
```diff
@@ -47,7 +47,8 @@ export class RichTextMarkdownSerializer {
             doc: defaultMarkdownSerializer.nodes.doc!,
             paragraph: defaultMarkdownSerializer.nodes.paragraph!,
-            text: defaultMarkdownSerializer.nodes.text!
+            text: defaultMarkdownSerializer.nodes.text!,
```
Just added `,` to `text` line, So it's showing changes. It is already used in previous implementations.

vikisekarNI commented on packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:
```diff
@@ -890,6 +1046,39 @@ describe('RichTextEditor', () => {
 
+    describe('`getMarkdown` with hard break backslashes should be same immediately after `setMarkdown`', () => {
+        const hardBreakMarkdownStrings: { name: string }[] = [
+            { name: '**bold**\\\n*Italics*' },
+            { name: '* list\\\n  hard break content\n\n* list' },
+            {
+                name: '* list\\\n  hard break content\n\n* list\n\n  * nested list\\\n    nested hard break content'
+            },
+            { name: '1. list\\\n   hard break content\n\n2. list' },
+            {
+                name: '1. list\\\n   hard break content\n\n2. list\n\n   1. nested list\\\n      nested hard break content'
+            }
+        ];
```
Consider having a value here to keep the test case title simple as mentioned in [other comment](https://github.com/ni/nimble/pull/1517#discussion_r1323261302)

vikisekarNI commented on packages/nimble-components/src/rich-text/models/tests/markdown-parser.spec.ts:
```diff
@@ -344,4 +344,59 @@ describe('Markdown parser', () => {
+        for (const value of markdownStringWithHardBreak) {
+            const specType = getSpecTypeByNamedList(value, focused, disabled);
+            specType(
+                `should render br tag between "${value.name}"`,
```
```suggestion
                `should render br tag with "${value.name}"`,
```

aagash-ni replied:
Updated

aagash-ni commented on change/@ni-nimble-components-90bda8e4-208e-4e0c-ae48-b3730a529c57.json:
```diff
@@ -0,0 +1,7 @@
+{
```
I have added `@tiptap/extension-hard-break` package for force line break and will have keyboard accessibility `Ctrl/Shift/Cmd + Enter`, So It needs to be updated in the spec file.

Whether we can update the spec file in this same PR or separate PR is required? @jattasNI @rajsite

m-akinc commented:
As a reminder, please create the PR in draft mode and let a buddy (me) review before publishing and adding owners.

m-akinc commented on packages/nimble-components/src/rich-text/editor/testing/rich-text-editor.pageobject.ts:
```diff
@@ -114,7 +126,16 @@ export class RichTextEditorPageObject {
             lastElement = lastElement?.lastElementChild;
         }
-        lastElement!.parentElement!.textContent = value;
+        const breakTags = lastElement!.parentElement!.querySelectorAll('br');
+        if (breakTags && breakTags.length > 1) {
+            const textNode = document.createTextNode(value);
+            lastElement!.parentElement!.insertBefore(
+                textNode,
+                lastElement as Node
+            );
+        } else {
+            lastElement!.parentElement!.textContent = value;
+        }
```
I'm confused by expected behavior of this function. In the old implementation (and the `else` case in the new implementation), setting the `textContent` of the `parentElement` overwrites anything that might have been there, including the `lastElement`, right? Like in the following situation, the new text would replace both existing strings, right?
![image](https://github.com/ni/nimble/assets/7282195/3cac1f65-6e75-4527-b078-afc5a54d4249)
 But if there are `<br>` elements this function appends instead? For simplicity and consistency, maybe this should always append:
```suggestion
        const textNode = document.createTextNode(value);
        lastElement!.parentElement!.appendChild(textNode);
```
It looks like any placeholder elements (e.g. `ProseMirror-trailingBreak`) get removed when you insert after them.

aagash-ni commented on packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:
```diff
@@ -912,6 +1068,47 @@ describe('RichTextEditor', () => {
 
+    describe('`getMarkdown` with hard break backslashes should be same immediately after `setMarkdown`', () => {
+        const hardBreakMarkdownStrings: { name: string, value: string }[] = [
+            { name: 'bold and italics', value: '**bold**\\\n*Italics*' },
```
In `String.raw` `\n` is considered as raw string, So in cases containing new lines, will look like the below image when we add `String.raw`. 

Still, we can add this if it's fine for you. Please let me know your thoughts.

![image](https://github.com/ni/nimble/assets/123377167/b37e611b-723a-45cd-bad6-d521adbce0b7)

m-akinc replied:
Let's go with that. At least then it's easier to see how whitespace affects alignment, and the backslashes stand out better.

aagash-ni replied:
Updated

aagash-ni commented on packages/nimble-components/src/rich-text/editor/tests/rich-text-editor.spec.ts:
```diff
@@ -418,6 +471,25 @@ describe('RichTextEditor', () => {
             ).toBeTrue();
         });
 
+        it('should have br tag name when pressing shift + Enter with nested numbered lists content', async () => {
```
Removed a few cases where the hard breaks are checked in a mixed nested list( Bulleted list with nested numbered list and vice versa ). I had other cases where I checked the first level of the nested list just to verify the DOM.

atmgrifter00 commented:
You can bypass me for this review.

rajsite reviewed (commented):
Waiting to see results of test changes

aagash-ni commented:
> Waiting to see results of test changes

Hi @rajsite , I hope you are waiting for the Mert test change comments to be addressed, please let us know if you have any other concerns. I have addressed all the comments that Mert mentioned.

rajsite commented:
Can bypass @mollykreis and @atmgrifter00, so ready to merge!
